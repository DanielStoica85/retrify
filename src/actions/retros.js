import database from '../config/firebase';

// Add retro
export const addRetro = retro => ({
    type: 'ADD_RETRO',
    retro
});

// add retro to db and then to store
export const startAddRetro = (retroData = {}) => {
    return dispatch => {
        const {
            title = '',
            description = '',
            createdAt = 0,
            author = ''
        } = retroData;

        const retro = { title, description, createdAt, author };

        return database
            .ref('retros')
            .push(retro)
            .then(ref => {
                dispatch(
                    addRetro({
                        id: ref.key,
                        ...retro
                    })
                );
            });
    };
};

// Remove retro
export const removeRetro = id => ({
    type: 'REMOVE_RETRO',
    id
});

// remove retro from db
export const startRemoveRetro = id => {
    return dispatch => {
        return database
            .ref(`retros/${id}`)
            .remove()
            .then(() => {
                dispatch(removeRetro(id));
            });
    };
};

// Edit retro
export const editRetro = (id, updates) => ({
    type: 'EDIT_RETRO',
    id,
    updates
});

// Start edit retro
export const startEditRetro = (id, updates) => {
    return dispatch => {
        return database
            .ref(`retros/${id}`)
            .update(updates)
            .then(() => {
                dispatch(editRetro(id, updates));
            });
    };
};

// Set retros
export const setRetros = retros => ({
    type: 'SET_RETROS',
    retros
});

// Start set retros
export const startSetRetros = () => {
    return dispatch => {
        return database
            .ref('retros')
            .once('value')
            .then(snapshot => {
                let retros = [];
                snapshot.forEach(element => {
                    retros.push({
                        id: element.key,
                        ...element.val()
                    });
                });
                dispatch(setRetros(retros));
            });
    };
};
