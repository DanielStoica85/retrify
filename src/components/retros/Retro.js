import React from 'react';
import {
    Card,
    CardBody,
    CardTitle,
    CardSubtitle,
    CardFooter
} from 'reactstrap';
import PropTypes from 'prop-types';

const Retro = props => {
    const { title, author, description, createdAt } = props.retro;
    return (
        <div>
            <Card className="mt-3">
                <CardBody>
                    <CardTitle>{title}</CardTitle>
                    <CardSubtitle> {description}</CardSubtitle>
                </CardBody>
                <CardFooter
                    style={{
                        backgroundColor: 'lightgray',
                        borderColor: 'lightgray'
                    }}
                >
                    Added by {author} on {createdAt}
                </CardFooter>
            </Card>
        </div>
    );
};

Retro.propTypes = {
    retro: PropTypes.object
};

export default Retro;
