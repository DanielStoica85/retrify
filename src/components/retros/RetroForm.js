import React, { Component } from 'react';
import LabeledFormInput from '../layout/LabeledFormInput';
import LabeledFormSelect from '../layout/LabeledFormSelect';
import { Button, Form, Label, FormGroup, Alert } from 'reactstrap';

import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

class RetroForm extends Component {
    constructor(props) {
        super(props);
        this.state = {
            title: props.retro ? props.retro.title : '',
            description: props.retro ? props.retro.description : '',
            author: props.retro ? props.retro.author : '',
            type: props.retro ? props.retro.type : '',
            createdAt: props.retro ? moment(props.retro.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    handleChange = e => {
        this.setState({
            [e.target.id]: e.target.value
        });
    };

    handleTypeChange = e => {
        const type = e.target.value;
        if (type !== 'Select') {
            this.setState({ type });
        } else {
            this.setState({ type: '' });
        }
    };

    handleDateChange = createdAt => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    handleFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    handleSubmit = e => {
        e.preventDefault();
        if (
            !this.state.title ||
            !this.state.author ||
            !this.state.createdAt ||
            !this.state.type
        ) {
            this.setState(() => ({
                error:
                    'Please provide a title, author name, retro type and a date for this retro.'
            }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                title: this.state.title,
                author: this.state.author,
                description: this.state.description,
                type: this.state.type,
                createdAt: this.state.createdAt.valueOf()
            });
        }
    };

    render() {
        return (
            <Form onSubmit={this.handleSubmit}>
                {this.state.error && (
                    <Alert color="danger">{this.state.error}</Alert>
                )}
                <LabeledFormInput
                    labelText="Retro Board Name*"
                    labelFor="title"
                    inputType="text"
                    inputName="title"
                    inputId="title"
                    value={this.state.title}
                    handleChange={this.handleChange}
                />
                <LabeledFormInput
                    labelText="Author Name*"
                    labelFor="author"
                    inputType="text"
                    inputName="author"
                    inputId="author"
                    value={this.state.author}
                    handleChange={this.handleChange}
                />
                <LabeledFormInput
                    labelText="Description"
                    labelFor="description"
                    inputType="textarea"
                    inputName="description"
                    inputId="description"
                    inputPlaceHolder="Give your board an (optional) description..."
                    value={this.state.description}
                    handleChange={this.handleChange}
                />
                <LabeledFormSelect
                    labelText="Type of Retrospective*"
                    labelFor="type"
                    inputType="select"
                    inputName="type"
                    inputId="type"
                    value={this.state.type}
                    options={[
                        'Select',
                        'What went well / What did not go well',
                        'Start / Stop / Continue'
                    ]}
                    handleChange={this.handleTypeChange}
                />
                <FormGroup>
                    <Label for="date-picker">Date*</Label>
                    <br />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.handleDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.handleFocusChange}
                        numberOfMonths={1}
                        id="date-picker"
                    />
                </FormGroup>
                <FormGroup>
                    <Button color="primary">Save Retro</Button>
                </FormGroup>
            </Form>
        );
    }
}

export default RetroForm;
