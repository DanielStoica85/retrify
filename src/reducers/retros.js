// Expenses reducer
const retrosReducerDefaultState = [];

export default (state = retrosReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_RETRO':
            return [...state, action.retro];
        case 'EDIT_RETRO':
            return state.map(retro => {
                if (retro.id === action.id) {
                    return {
                        ...retro,
                        ...action.updates
                    };
                } else {
                    return retro;
                }
            });
        case 'REMOVE_RETRO':
            return state.filter(retro => retro.id !== action.id);
        case 'SET_RETROS':
            return action.retros;
        default:
            return state;
    }
};
