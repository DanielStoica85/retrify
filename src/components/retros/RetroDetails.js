import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startRemoveRetro } from '../../actions/retros';
import {
    Container,
    Row,
    Col,
    Card,
    CardBody,
    CardTitle,
    CardText,
    CardFooter,
    Button
} from 'reactstrap';
import PropTypes from 'prop-types';

export class RetroDetails extends React.Component {
    onRemove = () => {
        this.props.startRemoveRetro(this.props.retro.id);
        this.props.history.push('/');
    };

    render() {
        const { title, description, author, createdAt } = this.props.retro;

        return (
            <Container className="retro-details">
                <Card>
                    <CardBody>
                        <CardTitle>{title}</CardTitle>
                        <CardText>{description}</CardText>
                    </CardBody>
                    <CardFooter
                        style={{
                            backgroundColor: 'lightgray',
                            borderColor: 'lightgray'
                        }}
                    >
                        <Row>
                            <Col md="8">
                                {' '}
                                <div>
                                    Posted by {author} on {createdAt}.
                                </div>
                            </Col>
                            <Col md="2">
                                <Link to="/">
                                    <Button
                                        className="float-right"
                                        color="info"
                                    >
                                        Back to retros list
                                    </Button>
                                </Link>
                            </Col>
                            <Col md="2">
                                <Button
                                    onClick={this.onRemove}
                                    className="float-left"
                                    color="danger"
                                >
                                    Remove retro
                                </Button>
                            </Col>
                        </Row>
                    </CardFooter>
                </Card>
            </Container>
        );
    }
}

const mapStateToProps = (state, props) => {
    const id = props.match.params.id;
    return {
        retro: state.retros.find(retro => retro.id.toString() === id)
    };
};

const mapDispatchToProps = (dispatch, props) => {
    return {
        startRemoveRetro: id => dispatch(startRemoveRetro(id))
    };
};

RetroDetails.propTypes = {
    match: PropTypes.object
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(RetroDetails);
