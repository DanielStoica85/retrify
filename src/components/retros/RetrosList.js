import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Retro from './Retro';

const RetrosList = props => {
    const { retros } = props;
    const retrosList =
        retros &&
        retros.map(retro => {
            return <Retro retro={retro} key={retro.id} />;
        });

    return (
        <div className="retros-list section">
            <p>Retros</p>
            {retros.length > 0 ? (
                <p>{retrosList}</p>
            ) : (
                <p>
                    You currently have no retros. Click{' '}
                    <Link to="/retro">HERE</Link> to add your first retro.
                </p>
            )}
        </div>
    );
};

RetrosList.propTypes = {
    retros: PropTypes.array
};

export default RetrosList;
