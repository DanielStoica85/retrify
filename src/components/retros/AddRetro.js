import React from 'react';
import { connect } from 'react-redux';
import { addRetro } from '../../actions/retros';
import { Container } from 'reactstrap';
import RetroForm from './RetroForm';

export class AddRetro extends React.Component {
    addRetro = retro => {
        this.props.addRetro(retro);
        this.props.history.push('/');
    };

    render() {
        return (
            <Container>
                <h4 className="grey-text text-darken-3">Add Retro Board</h4>
                <RetroForm onSubmit={this.addRetro} />
            </Container>
        );
    }
}

const mapDispatchToProps = dispatch => {
    return {
        addRetro: retro => dispatch(addRetro(retro))
    };
};

export default connect(
    undefined,
    mapDispatchToProps
)(AddRetro);
