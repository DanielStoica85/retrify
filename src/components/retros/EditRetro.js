import React from 'react';
import { connect } from 'react-redux';
import { startEditRetro } from '../../actions/retros';
import { Container } from 'reactstrap';
import RetroForm from './RetroForm';
import PropTypes from 'prop-types';

export class EditRetro extends React.Component {
    onSubmit = retro => {
        this.props.startEditRetro(this.props.retro.id, retro);
        this.props.history.push('/');
    };
    render() {
        return (
            <Container>
                <h4 className="grey-text text-darken-3">
                    Edit Retro Board with id {this.props.match.params.id}
                </h4>
                <RetroForm retro={this.props.retro} onSubmit={this.onSubmit} />
            </Container>
        );
    }
}

const mapDispatchToProps = (dispatch, props) => {
    return {
        startEditRetro: (id, retro) => dispatch(startEditRetro(id, retro))
    };
};

const mapStateToProps = (state, props) => {
    const id = props.match.params.id;
    return {
        retro: state.retros.find(retro => retro.id.toString() === id)
    };
};

EditRetro.propTypes = {
    startEditRetro: PropTypes.func
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(EditRetro);
