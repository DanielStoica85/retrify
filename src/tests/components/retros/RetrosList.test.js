import React from 'react';
import { shallow } from 'enzyme';
import RetrosList from '../../../components/retros/RetrosList';
import Retro from '../../../components/retros/Retro';
import retros from '../../fixtures/retros';

describe('RetrosList component', () => {
    it('should render correctly without retros', () => {
        const component = shallow(<RetrosList retros={[]} />);
        expect(component.find(Retro).length).toBe(0);
        expect(component).toMatchSnapshot();
    });

    it('should render correctly with retros', () => {
        const component = shallow(<RetrosList retros={retros} />);
        expect(component.find(Retro).length).toBe(retros.length);
        expect(component).toMatchSnapshot();
    });
});
