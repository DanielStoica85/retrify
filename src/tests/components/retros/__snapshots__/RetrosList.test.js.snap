// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RetrosList component should render correctly with retros 1`] = `
<div
  className="retros-list section"
>
  <p>
    Retros
  </p>
  <p>
    <Retro
      key="1"
      retro={
        Object {
          "author": "Gica",
          "createdAt": 1564169537000,
          "description": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Deleniti itaque aut eligendi laborum excepturi animi hic similique impedit enim velit.",
          "id": "1",
          "title": "First Retro",
        }
      }
    />
    <Retro
      key="2"
      retro={
        Object {
          "author": "Costel",
          "createdAt": 1564255937000,
          "description": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Deleniti itaque aut eligendi laborum excepturi animi hic similique impedit enim velit.",
          "id": "2",
          "title": "Second Retro",
        }
      }
    />
    <Retro
      key="3"
      retro={
        Object {
          "author": "Miron",
          "createdAt": 1564342337000,
          "description": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Deleniti itaque aut eligendi laborum excepturi animi hic similique impedit enim velit.",
          "id": "3",
          "title": "Third Retro",
        }
      }
    />
  </p>
</div>
`;

exports[`RetrosList component should render correctly without retros 1`] = `
<div
  className="retros-list section"
>
  <p>
    Retros
  </p>
  <p>
    You currently have no retros. Click
     
    <Link
      to="/retro"
    >
      HERE
    </Link>
     to add your first retro.
  </p>
</div>
`;
